File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.cpp
#include <string>
#include <vector>

using namespace std;

int one=0, zero=0;

void Check(int x, int y, int size, vector<vector<int>> &arr) {
    if(size == 0)
        return;
    
    bool zero_check = true, one_check = true;
    
    for(int i=y; i<y+size; i++){
        for(int j=x; j<x+size; j++){
            if(arr[i][j] == 1)
                zero_check = false;
            if(arr[i][j] == 0)
                one_check = false;
            
            if(!(zero_check || one_check))
                break;
        }
        
        if(!(zero_check || one_check))
            break;
    }
    
    if(zero_check){
        zero++;
        return;
    }
    if(one_check){
        one++;
        return;
    }
    
    Check(x, y, size/2, arr);
    Check(x+size/2, y, size/2, arr);
    Check(x, y+size/2, size/2, arr);
    Check(x+size/2, y+size/2, size/2, arr);
}

vector<int> solution(vector<vector<int>> arr) {
    vector<int> answer;
    
    Check(0, 0, arr.size(), arr);
    answer.push_back(zero);
    answer.push_back(one);
    return answer;
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.js
function solution(arr) {
    let answer = [];
    
    let zero=0, one=0;
    
    function Check(x, y, size){
        if(size === 0)
            return;
        
        let zeroCheck=true;
        let oneCheck=true;
        
        for(let i=y; i<y+size; i++){
            for(let j=x; j<x+size; j++){
                if(arr[i][j] === 0)
                    oneCheck = false;
                if(arr[i][j] === 1)
                    zeroCheck = false;
                
                if(!(oneCheck || zeroCheck))
                    break;
            }
            
            if(!(oneCheck || zeroCheck))
                break;
        }
        
        if(oneCheck){
            one++;
            return;
        }
        if(zeroCheck){
            zero++;
            return;
        }
        
        Check(x, y, size/2);
        Check(x+size/2, y, size/2);
        Check(x, y+size/2, size/2);
        Check(x+size/2, y+size/2, size/2);
    }
    
    Check(0, 0, arr.length, arr);
    answer.push(zero);
    answer.push(one);
    return answer;
}